pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,317 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/davecgh/go-spew/spew"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	extauthzhttp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/ext_authz/v3"
+	extauthztcp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/ext_authz/v3"
+	envoy_type_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	envoytypev3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/hashicorp/go-multierror"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	"istio.io/istio/pkg/config/host"
+)
+
+const (
+	extAuthzMatchPrefix = "istio-ext-authz"
+)
+
+var (
+	rbacDefaultDenyAll = &rbacpb.RBAC{
+		Action: rbacpb.RBAC_DENY,
+		Policies: map[string]*rbacpb.Policy{
+			"default-deny-all-due-to-bad-CUSTOM-action": {
+				Permissions: []*rbacpb.Permission{{Rule: &rbacpb.Permission_Any{Any: true}}},
+				Principals:  []*rbacpb.Principal{{Identifier: &rbacpb.Principal_Any{Any: true}}},
+			},
+		},
+	}
+)
+
+type builtExtAuthz struct {
+	http *extauthzhttp.ExtAuthz
+	tcp  *extauthztcp.ExtAuthz
+}
+
+func notAllTheSame(names []string) bool {
+	for i := 1; i < len(names); i++ {
+		if names[i-1] != names[i] {
+			return true
+		}
+	}
+	return false
+}
+
+func buildExtAuthz(configs []*meshconfig.MeshConfig_ExtensionProvider, providers []string) (*builtExtAuthz, error) {
+	if notAllTheSame(providers) {
+		return nil, fmt.Errorf("all extension providers must be the same for a specific workload, found multiple different providers: %v", providers)
+	} else if len(providers) < 1 {
+		return nil, fmt.Errorf("no extension provider found")
+	}
+	provider := providers[0]
+
+	resolved := map[string]*builtExtAuthz{}
+	var errs error
+	if len(configs) == 0 {
+		errs = multierror.Append(errs, fmt.Errorf("at least 1 extension provider must be defined"))
+	}
+	for i, config := range configs {
+		if config.Name == "" {
+			errs = multierror.Append(errs, fmt.Errorf("extension provider name must not be empty, found empty at index: %d", i))
+		} else if _, found := resolved[config.Name]; found {
+			errs = multierror.Append(errs, fmt.Errorf("extension provider name must be unique, found duplicate: %s", config.Name))
+		}
+		var parsed *builtExtAuthz
+		var err error
+		switch p := config.Provider.(type) {
+		case *meshconfig.MeshConfig_ExtensionProvider_EnvoyExtAuthzHttp:
+			parsed, err = buildExtAuthzHTTP(p.EnvoyExtAuthzHttp)
```

Comment:
Can we cache the result

Reply:
probably in a follow PR, there might still some big changes to the implementation here, added a TODO.